catalog/service/rest/src/main/java/org/projectnessie/catalog/service/rest/IcebergApiV1GenericResource.java

```diff
@@ -175,4 +179,129 @@ public Uni<Void> commitTransaction(
+          .call(this::checkForbiddenLocations)
+          .flatMap(this::fetchSnapshot)
+          .onItem()
+          .transformToMulti(this::collectLocations)
+          .filter(requestedS3Uri::startsWith)
+          .toUni()
+          .onItem()
+          .ifNull()
+          .failWith(this::unauthorized)
+          .replaceWith(request.uri())
+          .map(this::sign);
+    }
+
+    private Uni<Content> fetchContent() {
+      try {
+        ContentResponse contentResponse =
+            nessieApi
+                .getContent()
+                .refName(ref.name())
+                .hashOnRef(ref.hashWithRelativeSpec())
+                .getSingle(key);
+        return Uni.createFrom().item(contentResponse.getContent());
+      } catch (NessieContentNotFoundException e) {
+        return Uni.createFrom().nullItem();
+      } catch (NessieNotFoundException e) {
+        return Uni.createFrom().failure(e);
+      }
+    }
+
+    private Uni<?> checkForbiddenLocations(Content content) {
+      String metadataLocation = null;
+      if (content instanceof IcebergTable) {
+        metadataLocation = ((IcebergTable) content).getMetadataLocation();
+      } else if (content instanceof IcebergView) {
+        metadataLocation = ((IcebergView) content).getMetadataLocation();
+      }
+      if (metadataLocation == null || !request.uri().equals(metadataLocation)) {
+        return Uni.createFrom().nullItem();
+      } else {
+        return Uni.createFrom().failure(unauthorized());
+      }
+    }
+
+    private Uni<NessieEntitySnapshot<?>> fetchSnapshot(Content content) {
+      if (content != null) {
+        try {
+          CompletionStage<SnapshotResponse> stage =
+              catalogService.retrieveSnapshot(
+                  SnapshotReqParams.forSnapshotHttpReq(ref, "iceberg", null),
+                  key,
+                  content.getType());
+          return Uni.createFrom().completionStage(stage).map(SnapshotResponse::nessieSnapshot);
+        } catch (NessieContentNotFoundException ignored) {
+        } catch (NessieNotFoundException e) {
+          return Uni.createFrom().failure(e);
+        }
+      }
+      return Uni.createFrom().nullItem();
+    }
+
+    private Multi<String> collectLocations(NessieEntitySnapshot<?> snapshot) {
+      if (snapshot != null) {
+        // table exists: collect all locations
+        return Multi.createFrom()
+            .<String>emitter(
+                e -> {
+                  e.emit(snapshot.icebergLocation());
+                  snapshot.additionalKnownLocations().forEach(e::emit);
+                  e.complete();
+                })
+            // if location is within the base prefix, use the base prefix;
+            // otherwise, use the location as-is. This allows for temporary
+            // locations to be used for signing, as long as they are not
+            // outside the base prefix.
+            .map(location -> location.startsWith(basePrefix) ? basePrefix : location);
```

Comment:
Hm - not really a fan, because it basically exposes the whole basePrefix. Is this only for one or two tests?

If it's more than that, I'd be okay to leave it as it is but have a follow-up issue to tackle this now.

Reply:
Yes it's basically to make tests pass. Would you be OK if we changed the tests, or disabled them?

Reply:
I think, we can live without this fallback.

`testRegisterTableFromFileSystem` ... maybe let that write the metadata of the table to be registered somewhere underneath the original table's location - dirty, but at least a quick fix.
Side note: the javadoc of that one needs to be changed.

`testCompleteCreateOrReplaceTransactionReplace` + `testCompleteReplaceTransaction` could write the file directly into the table's location directory (somewhere in there at least).

WDYT?

Reply:
That wouldn't work, the tests do not know the definitive table location upfront, because it contains a random uuid.

At this point, I think the only option is do disable the tests, or change them in a way that they use the default location, not a custom one.

Reply:
The location should be there from the create-table before.

Reply:
In certain tests there is no create-table before:

https://github.com/projectnessie/nessie/blob/170dd11d4c04dbccc042b7aa11cb76f9fa144c17/servers/quarkus-server/src/testFixtures/java/org/projectnessie/server/catalog/CatalogTests.java#L630-L637

Reply:
But that can be "guarded" with an `if (overridesRequestedLocation())` and omit the `withLocation`